distributed/dashboard/components/scheduler.py

```diff
@@ -1720,6 +1724,214 @@ def __del__(self):
+        self.x = {}
+        self.y = {}
+        self.collision = {}
+        self.y_next = 0
+        self.nstart = []
+        self.nend = []
+        self.tg_stack = []
+
+        self.node_source = ColumnDataSource(
+            {
+                "x": [],
+                "y": [],
+                "name": [],
+                "name_short": [],
+                "tot_tasks": [],
+                "colors": [],
+            }
+        )
+
+        self.arrow_source = ColumnDataSource({"xs": [], "ys": [], "xe": [], "ye": []})
+
+        self.root = figure(title="Task Groups Graph", **kwargs)
+        self.subtitle = Title(text=" ", text_font_style="italic")
+        self.root.add_layout(self.subtitle, "above")
+
+        # Let's start with squares we, will modify later to progress bars
+        rect = self.root.square(
+            x="x",
+            y="y",
+            size=30,
+            color="colors",
+            fill_alpha=0.5,
+            source=self.node_source,
+        )
+
+        self.arrows = Arrow(
+            end=VeeHead(size=10),
+            line_color="black",
+            line_alpha=0.5,
+            line_width=1,
+            x_start="xs",
+            y_start="ys",
+            x_end="xe",
+            y_end="ye",
+            source=self.arrow_source,
+        )
+        self.root.add_layout(self.arrows)
+
+        self.labels = LabelSet(
+            x="x",
+            y="y",
+            text="name_short",
+            source=self.node_source,
+            background_fill_color=None,
+        )  #  We probably need an offset like x_offset=-1, also need to chop the name somehow
+        self.root.add_layout(self.labels)
+
+        self.root.xgrid.grid_line_color = None
+        self.root.ygrid.grid_line_color = None
+
+        hover = HoverTool(
+            point_policy="follow_mouse",
+            tooltips=[("tg", "@name"), ("num_task", "@tot_tasks")],
+            renderers=[rect],
+        )
+        self.root.add_tools(hover)
+
+    @without_property_validation
+    def update(self):
+
+        with log_errors():
+
+            if self.scheduler.task_groups:
+                # get dependecies per task group
+                dependencies = {
+                    k: [
+                        ds.name for ds in ts.dependencies if ds.name != k
+                    ]  # in some cases there are tg that have themeselves as
+                    for k, ts in self.scheduler.task_groups.items()  # dependencies, we remove those.
+                }
+
+                # get dependents per task group
+                dependents = {k: [] for k in dependencies}
+                for k, v in dependencies.items():
+                    for dep in v:
+                        dependents[dep].append(k)
+
+                dependencies_depth = {
+                    k: get_depth(dependencies, k) for k in dependencies.keys()
+                }
+
+                stack_order = toposort_layers(dependencies)
```

Comment:
If we're toposorting then we may not need to call `get_depth`.  We can just call something like the following:

```python
depth[tg] = max(depth[dep] for dep in tg.dependencies) + 1`
```

Reply:
Good point, changed the code to remove the use of get_depth.